adding proxy_address field for active proxies

rockaut opened the pull request:
##### SUMMARY

Since 4.0.0 there is a field proxy_address for limiting proxy requests.

Official Docs: "Proxy address	If specified then active proxy requests are only accepted from this list of comma-delimited IP addresses, optionally in CIDR notation, or DNS names of active Zabbix proxy.
This field is only available if an active proxy is selected in the Proxy mode field. Macros are not supported.
This option is supported since Zabbix 4.0.0."

##### ISSUE TYPE
- Feature Pull Request

##### COMPONENT NAME
zabbix_proxy

##### ADDITIONAL INFORMATION
```yaml
---

- hosts:
    - localhost
  tasks:
    - name: Create a new proxy or update an existing active proxy
      local_action:
        module: zabbix_proxy
        server_url: "{{ server_url }}"
        login_user: "{{ login_user }}"
        login_password: "{{ login_password }}"
        proxy_name: ExampleProxy
        description: ExampleProxy
        status: active
        state: present
        proxy_address: ExampleProxy
    - name: Create a new proxy or update an existing passive proxy
      local_action:
        module: zabbix_proxy
        server_url: "{{ server_url }}"
        login_user: "{{ login_user }}"
        login_password: "{{ login_password }}"
        proxy_name: ExampleProxy2
        description: ExampleProxy2
        status: passive
        state: present
        proxy_address: ExampleProxy2
        interface:
          type: 0
          main: 1
          useip: 1
          ip: 10.0.0.1
          dns: "example.local"
          port: 10050
```
ExampleProxy is active and will be created with proxy_address field. ExampleProxy2 is passive and will be created without proxy_address field but with interface.

Both fields are marked as not required. Depending on the status [active|passive] the corresponding field is kept or poped.
Zabbix Version lower than 4.0 can still be used if proxy_address is just not defined in config.

D3DeFi commented on lib/ansible/modules/monitoring/zabbix/zabbix_proxy.py:
```diff
@@ -36,12 +36,18 @@
 requirements:
     - "python >= 2.6"
-    - "zabbix-api >= 0.5.4"
+    - "zabbix-api >= 0.5.3"
```
Is there any reason for this change? It will break module as is right now. All zabbix modules are now requiring `zabbix-api=0.5.4` as version 0.5.3 didn't support `zbx.logout()` method, which resulted in session being left as active on the zabbix server

rockaut replied:
no reason - will change

D3DeFi commented on lib/ansible/modules/monitoring/zabbix/zabbix_proxy.py:
```diff
@@ -165,6 +172,16 @@ def add_proxy(self, data):
+                parameters.pop('interface')
+            if 'interface' in data and data['status'] == '6':
+                parameters['interface'] = data['interface']
+
```
Is there a reason to handling this differently than changes you've introduced in `update_proxy()` at lines 228-232?

rockaut replied:
yes - if handled the same way ansible result was always "changed". This way it works properly if nothing is changed.

D3DeFi replied:
Not required to change, but something like this seems to be working for me in the same way:

```python
            if 'proxy_address' in data and data['status'] != '5':
                parameters.pop('proxy_address', False)

            if 'interface' in data and data['status'] != '6':
                parameters.pop('interface', False)

```

D3DeFi reviewed (changes requested):
Thanks for the patch! :)

Too bad this argument is not [documented](https://www.zabbix.com/documentation/current/manual/api/reference/proxy/object) in Zabbix API documentation.

D3DeFi commented on lib/ansible/modules/monitoring/zabbix/zabbix_proxy.py:
```diff
@@ -36,12 +36,18 @@
+    proxy_address:
+        description:
+            - Address of the active proxy in Zabbix to limit access ( Zabbix Version >= 4.0.0 )
+            - Remove option for Zabbix Version < 4.0.0
```
This should be documented more clearly , I would propose something like this to align with other zabbix modules:

```yml
        description:
            - Comma-delimited list of IP/CIDR addresses or DNS names to accept active proxy requests from.
            - Requires I(status=active).
            - Works only with >= Zabbix 4.0.
```

rockaut replied:
Looks good - will change. I will just additionally mention the option remove:

`- Works only with >= Zabbix 4.0. ( remove option for <= 4.0 )`

rockaut commented:
You're right @D3DeFi - sadly most of the API properties today aren't documented but productive none-the-less.

rockaut commented:
I created an Issue for the missing documentation here: https://support.zabbix.com/browse/ZBX-17154

sky-joker commented:
Hi @rockaut

Thank you for the patch!
Could you please include a changelog fragment file in this PR?
https://docs.ansible.com/ansible/latest/community/development_process.html#changelogs-how-to

D3DeFi reviewed (commented):
Just finished testing against Zabbix 4.0 and works well. Passing integration tests introduced (not yet merged) in #66033 too + some additional tests.

@sky-joker can you test this as well please? As you've mentioned, changelog fragment should be part of this PR.

rockaut commented:
Have added changelog fragment and also changed to your suggestions @D3DeFi, thank you. I think i just mixed the status fields yesterday. Works now on 4.4

D3DeFi reviewed (approved):
Looks good to me now :) Thanks for implementing my suggestions.

Maybe it will be worth squashing commits into a single one, but lets wait for @sky-joker 's review first

sky-joker commented:
Looks good :)
Could you please squash commits?

rockaut committed:
adding proxy_address field for active proxies

Since 4.0.0 there is a field proxy_address for limiting proxy requests.

Official Docs: "Proxy address	If specified then active proxy requests are only accepted from this list of comma-delimited IP addresses, optionally in CIDR notation, or DNS names of active Zabbix proxy.
This field is only available if an active proxy is selected in the Proxy mode field. Macros are not supported.
This option is supported since Zabbix 4.0.0."
updated documentation/help
fixed test issues
fixed version_added not a string
fix suggestions by D3DeFi


added changelog fragment
applied D3DeFi suggestions

way cleaner and better readable

D3DeFi reviewed (approved):
Thank you all for all the work! 
shipit

sky-joker commented:
shipit
